File: articles.jdbc/src/test/java/pet/db/jdbc/tool/generator/UserPayloadTestDataGenerator.java
package pet.db.jdbc.tool.generator;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Component;

import pet.db.jdbc.model.dto.payload.UserPayload;
import pet.db.jdbc.model.dto.User;

import java.util.List;

@Component
@RequiredArgsConstructor
public class UserPayloadTestDataGenerator implements TestDataGenerator<UserPayload> {

    private final TestDataGenerator<User> userTestDataGenerator;

    @Override
    public UserPayload generateSavedData() {
        return generateSavedData(1).getFirst();
    }

    @Override
    public UserPayload generateUnsavedData() {
        return generateUnsavedData(1).getFirst();
    }

    @Override
    public List<UserPayload> generateSavedData(Integer dataSize) {
        return userTestDataGenerator.generateSavedData(dataSize).stream()
                .map(this::convertToUserPayload)
                .toList();
    }

    @Override
    public List<UserPayload> generateUnsavedData(Integer dataSize) {
        return userTestDataGenerator.generateUnsavedData(dataSize).stream()
                .map(this::convertToUserPayload)
                .toList();
    }

    private UserPayload convertToUserPayload(User user) {
        return new UserPayload(
                user.getUsername(),
                user.getEmail(),
                user.getPassword(),
                user.getRole());
    }

}

File: src/test/kotlin/pet/articles/test/tool/generator/UserPayloadTestDataGenerator.kt
package pet.articles.test.tool.generator

import org.koin.core.annotation.Named
import org.koin.core.annotation.Single

import pet.articles.model.dto.User
import pet.articles.model.dto.payload.UserPayload

class UserPayloadTestDataGenerator(
    private val userTestDataGenerator: TestDataGenerator<User>
) : TestDataGenerator<UserPayload> {

    override fun generateUnsavedData(dataSize: Int): List<UserPayload> =
        userTestDataGenerator.generateUnsavedData(dataSize).map(::convertToUserPayload)

    override fun generateSavedData(dataSize: Int): List<UserPayload> =
        userTestDataGenerator.generateSavedData(dataSize).map(::convertToUserPayload)

    override fun generateInvalidData(): UserPayload =
        convertToUserPayload(userTestDataGenerator.generateInvalidData())

    private fun convertToUserPayload(user: User): UserPayload =
        UserPayload(
            username = user.username,
            email = user.email,
            password = user.password!!,
            role = user.role
        )
}